src/sql/workbench/services/contextualization/common/serverContextualizationService.ts

```diff
@@ -63,32 +56,86 @@ export class ServerContextualizationService extends Disposable implements IServe
 		throw invalidProvider(providerId);
 	}
 
+	/**
+	 * Contextualizes the provided URI for GitHub Copilot.
+	 * @param uri The URI to contextualize for Copilot.
+	 * @returns Copilot will have the URI contextualized when the promise completes.
+	 */
+	public async contextualizeUriForCopilot(uri: string): Promise<void> {
+		// Don't need to take any actions if contextualization is not enabled and can return
+		const isContextualizationNeeded = await this.isContextualizationNeeded();
+		if (!isContextualizationNeeded) {
+			return;
+		}
+
+		const getServerContextualizationResult = await this.getServerContextualization(uri);
+		if (getServerContextualizationResult.context) {
+			await this.sendServerContextualizationToCopilot(getServerContextualizationResult.context);
+		}
+		else {
+			const generateServerContextualizationResult = await this.generateServerContextualization(uri);
+			if (generateServerContextualizationResult.context) {
+				await this.sendServerContextualizationToCopilot(generateServerContextualizationResult.context);
+			}
+		}
+	}
+
 	/**
 	 * Generates server context
 	 * @param ownerUri The URI of the connection to generate context for.
 	 */
-	public generateServerContextualization(ownerUri: string): void {
+	private async generateServerContextualization(ownerUri: string): Promise<azdata.contextualization.GenerateServerContextualizationResult> {
 		const providerName = this._connectionManagementService.getProviderIdFromUri(ownerUri);
 		const handler = this.getProvider(providerName);
 		if (handler) {
-			handler.generateServerContextualization(ownerUri);
+			return await handler.generateServerContextualization(ownerUri);
+		}
+		else {
+			return Promise.resolve({
+				context: undefined
+			});
 		}
 	}
 
 	/**
 	 * Gets all database context.
 	 * @param ownerUri The URI of the connection to get context for.
 	 */
-	public async getServerContextualization(ownerUri: string): Promise<azdata.contextualization.GetServerContextualizationResult> {
+	private async getServerContextualization(ownerUri: string): Promise<azdata.contextualization.GetServerContextualizationResult> {
 		const providerName = this._connectionManagementService.getProviderIdFromUri(ownerUri);
 		const handler = this.getProvider(providerName);
 		if (handler) {
 			return await handler.getServerContextualization(ownerUri);
 		}
 		else {
 			return Promise.resolve({
```

Comment:
would it be helpful to also log an error here that the provider wasn't able to be retrieved for debugging?

Reply:
I'm going to have a follow up PR with all the logging.